pkg/ebpf/ingress_node_firewall_loader.go

```diff
@@ -55,43 +59,295 @@ func NewIngNodeFwController() (*IngNodeFwController, error) {
+// For each provided interface name:
+// i)   Look up the network interface by name.
+// ii)  Attach the program to the interface.
+// iii) Pin the XDP program.
+func (infc *IngNodeFwController) IngressNodeFwAttach(ifacesName ...string) error {
 	objs := infc.objs
-	rules := BpfRulesValSt{}
+	for _, ifaceName := range ifacesName {
+		if _, ok := infc.links[ifaceName]; ok {
+			klog.Infof("Interface %s is already attached and managed, skipping", ifaceName)
+			return nil
+		}
+		// Look up the network interface by name.
+		iface, err := net.InterfaceByName(ifaceName)
+		if err != nil {
+			return fmt.Errorf("lookup network iface %q: %s", ifaceName, err)
+		}
+
+		// Attach the program.
+		l, err := link.AttachXDP(link.XDPOptions{
+			Program:   objs.IngressNodeFirewallProcess,
+			Interface: iface.Index,
+		})
+		if err != nil {
+			if strings.Contains(err.Error(), xdpEBUSYErr) {
+				return fmt.Errorf("Interface %s is already attached", ifaceName)
+			}
+			return fmt.Errorf("could not attach XDP program: %s", err)
+		}
+		// Pin the XDP program.
+		lPinDir := path.Join(infc.pinPath, ifaceName+"_link")
+		if err := l.Pin(lPinDir); err != nil {
+			return fmt.Errorf("failed to pin link to pinDir %s: %s", lPinDir, err)
+		}
+		infc.links[ifaceName] = l
+		log.Printf("Attached IngressNode Firewall program to iface %q (index %d)", iface.Name, iface.Index)
+	}
+	return nil
+}
+
+// IngressNodeFwDetach detaches the eBPF program from the list of interfaces and cleans up the interfaces.
+// Additionally, it unloads all firewall rules that are associated to the interfaces.
+func (infc *IngNodeFwController) IngressNodeFwDetach(interfaceNames ...string) error {
+	// Detach from interfaces.
+	for _, interfaceName := range interfaceNames {
+		log.Printf("Detaching IngressNode Firewall program from interface %q", interfaceName)
+		if err := infc.cleanup(interfaceName); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// PurgeStaleInterfacesFromRules goes through all existing rules and purges keys for interfaces that no longer
+// exist. It first gets the list of stale interface keys (= keys for interfaces that are not part of the current
+// list of managed links). It then goes through the current ruleset and deletes all of these keys.
+func (infc *IngNodeFwController) PurgeStaleInterfacesFromRules() error {
+	// Get stale keys (= keys inside the eBPF map that do not match any managed interfaces).
+	staleKeys, err := infc.getStaleInterfaceKeys()
+	if err != nil {
+		return err
+	}
+
+	// Delete all keys that should be deleted.
+	if err := infc.purgeKeys(staleKeys); err != nil {
+		klog.Infof("Purge keys operation encountered issues, err: %q", err)
+	}
+
+	return nil
+}
+
+// GetBPFMapContent lists all existing keys and rules inside the map. Used for unit testing.
+func (infc *IngNodeFwController) GetBPFMapContent() (map[BpfLpmIpKeySt]BpfRulesValSt, error) {
```

Comment:
Can u add `Test` to the name so we are sure this only for unit-test

Reply:
done